File: IntegrationTests.VB/Parser/TypeParsing/BlacklistManagerIntegrationSample.vb
' Copyright 2013 rubicon informationstechnologie gmbh
'
' Licensed under the Apache License, Version 2.0 (the "License");
' you may not use this file except in compliance with the License.
' You may obtain a copy of the License at
'
' http://www.apache.org/licenses/LICENSE-2.0
'
' Unless required by applicable law or agreed to in writing, software
' distributed under the License is distributed on an "AS IS" BASIS,
' WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
' See the License for the specific language governing permissions and
' limitations under the License.
Imports InjectionCop.Attributes
Imports System
Imports System.Data
Imports System.Data.SqlClient

Namespace Parser.TypeParsing
	Public Class BlacklistManagerIntegrationSample
		Inherits ParserSampleBase

		Public Sub UnsafeBlacklistedCall()
			Dim command As IDbCommand = New SqlCommand()
			command.CommandText = MyBase.UnsafeSource("")
		End Sub

		Public Sub SafeBlacklistedCall()
			Dim command As IDbCommand = New SqlCommand()
			command.CommandText = MyBase.SafeSource()
		End Sub

		Public Sub ListedAndUnlistedViolation()
			MyBase.RequiresSqlFragment(MyBase.UnsafeSource())
			Dim command As IDbCommand = New SqlCommand()
			command.CommandText = MyBase.UnsafeSource("")
		End Sub

		Public Sub FragmentDefinedInXmlSafeCall()
			Me.FragmentDefinedInXml(Me.ReturnsExternal())
		End Sub

		Public Sub FragmentDefinedInXmlUnsafeCall()
			Me.FragmentDefinedInXml(Me.UnsafeString())
		End Sub

		Public Sub MixedViolations(x As Integer)
			Me.RequiresDummyFragment(3)
			Me.RequiresDummyFragment(Me.SafeDummySource())
			Me.RequiresDummyFragment(Me.UnsafeDummySource())
			Me.RequiresDummyFragment(Me.OtherUnsafeDummySource())
			Dim command As IDbCommand = New SqlCommand()
			command.CommandText = Me.UnsafeString()
			Me.FragmentDefinedInXml(Me.UnsafeString())
			Me.FragmentDefinedInXml(Me.ReturnsExternal())
			MyBase.RequiresSqlFragment("safe")
      If 1 = x Then
        If 2 = x Then
        End If
      End If
		End Sub

		Public Sub RequiresDummyFragment(<Fragment("dummy")> i As Integer)
		End Sub

		Public Function UnsafeDummySource() As Integer
      Return 3
		End Function

		Public Function OtherUnsafeDummySource() As<Fragment("wrongtype")>
		Integer
      Return 3
		End Function

		Public Function SafeDummySource() As<Fragment("dummy")>
		Integer
      Return 3
		End Function

		Public Function UnsafeString() As String
			Return"unsafe command"
		End Function

		Public Function ReturnsExternal() As<Fragment("external")>
		String
			Return"safe"
		End Function

		Public Sub FragmentDefinedInXml(needsExternalFragment As String)
		End Sub
	End Class
End Namespace

File: IntegrationTests/Parser/TypeParsing/BlacklistManagerIntegrationSample.cs
// Copyright 2012 rubicon informationstechnologie gmbh
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using System.Data;
using System.Data.SqlClient;
using InjectionCop.Attributes;

namespace InjectionCop.IntegrationTests.Parser.TypeParsing
{
  public class BlacklistManagerIntegrationSample : ParserSampleBase
  {
    public void UnsafeBlacklistedCall ()
    {
      IDbCommand command = new SqlCommand();
      command.CommandText = UnsafeSource ("");
    }

    public void SafeBlacklistedCall ()
    {
      IDbCommand command = new SqlCommand();
      command.CommandText = SafeSource();
    }

    public void ListedAndUnlistedViolation ()
    {
      RequiresSqlFragment (UnsafeSource());
      IDbCommand command = new SqlCommand();
      command.CommandText = UnsafeSource ("");
    }

    public void FragmentDefinedInXmlSafeCall ()
    {
      FragmentDefinedInXml (ReturnsExternal());
    }

    public void FragmentDefinedInXmlUnsafeCall ()
    {
      FragmentDefinedInXml (UnsafeString());
    }

    public void MixedViolations (int x)
    {
      RequiresDummyFragment (3);
      RequiresDummyFragment (SafeDummySource());
      RequiresDummyFragment (UnsafeDummySource());
      RequiresDummyFragment(OtherUnsafeDummySource());
      IDbCommand command = new SqlCommand();
      command.CommandText = UnsafeString();

      FragmentDefinedInXml (UnsafeString());
      FragmentDefinedInXml (ReturnsExternal());

      RequiresSqlFragment ("safe");             // safe since literal

      if (1 == x)
      {
        if (2 == x)
        {
          //requiresFragment (OtherUnsafeSource());
        }
        else
        {
          //requiresFragment (OtherUnsafeSource());
        }
      }
    }

    public void RequiresDummyFragment ([Fragment ("dummy")] int i)
    {
    }
    
    public int UnsafeDummySource ()
    {
      return 3;
    }

    [return:Fragment("wrongtype")]
    public int OtherUnsafeDummySource ()
    {
      return 3;
    }

    [return:Fragment("dummy")]
    public int SafeDummySource ()
    {
      return 3;
    }

    public string UnsafeString ()
    {
      return "unsafe command";
    }

    [return: Fragment ("external")]
    public string ReturnsExternal ()
    {
      return "safe";
    }

    public void FragmentDefinedInXml (string needsExternalFragment)
    {
    }
  }
}
